src/ptolemy/raster.py

```diff
@@ -561,37 +563,39 @@ def raster_to_df(raster, idxraster, idxmap, func=None, nodata=-1):
     Parameters
     ----------
     raster : xr.DataArray
-        an array of data aligned with the lat/lon dimensions of `idxraster`
+        data to make a pd.DataFrame
     idxraster : xr.DataArray
         an index raster, e.g., from `pt.Rasterize()`
-    idx_map : map
-        a map of strings to values of `idxraster` to generate the raster
-    func : unary function, optional, default: np.unique
-        a function with can be applied to an array of data
-    nodata : optional
-        the nodata value of `idxraster`
+    idx_map : dict, optional
+        a map of strings to values of `idxraster` if `idxraster` is not weighted
+    idx_dim : str, optional
+        the name of the index dimension if `idx_map` is provided
+    func : string, optional
+        a function with can be applied to an array of data. currently supports:
+            - max
+            - sum
+            - mean
+    drop_zeros : bool, optional
+        drop zeros from the dataframe before returning
     """
+    if idx_map:
+        idxraster = xr.concat(
+            [
+                xr.where(idxraster == v, 1, np.nan).expand_dims({idx_dim: [k]}, axis=1)
+                for k, v in idx_map.items()
+            ],
+            dim=idx_dim,
+        )
 
-    def default_func(ary):
-        v = np.unique(ary[~np.isnan(ary)])
-        if len(v) > 1:
-            raise ValueError("Non-unique values found in raster")
-        if len(v) == 0:
-            return np.nan
-        else:
-            return v[0]
-
-    if raster.shape != idxraster.shape:
-        raise ValueError("Raster and index raster must have same shape")
-
-    func = func or default_func
-    idxs = np.unique(idxraster)
-    idxs = idxs[(idxs != nodata) & (~np.isnan(idxs))]
-    data = {}
-    for idx in idxs:
-        ary = raster.values[idxraster.values == idx]
-        try:
-            data[idxmap[str(idx)]] = func(ary)
-        except:  # noqa: 722
-            continue
-    return pd.Series(data)
+    data = raster * idxraster
+    # TODO: there has to be a better way to do this
```

Comment:
any thoughts on how to do this better would be much appreciated. I trawled the xarray docs and didn't get much of anywhere..